### PR TITLE
Sync cg and statusDB

### PR DESCRIPTION
**Pull request by islean**
## Description

Addresses issue #2098. Main differences will be that sample.beaconized_at is dropped and the adding of some foreign keys to the actual db, where the Python models indicate that it should be so.

### Added

- Revision

### How to prepare for test

- [x] Ssh to relevant server (depending on type of change)
- [x] Use stage: `us`
- [x] Paxa the environment: `paxa`
- [x] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_main -t cg -b sync-cg-and-statusdb -a
    ```

### How to test

- [ ] Do ...

### Expected test outcome

- [ ] Check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [x] Tests executed by IO
- [x] "Merge and deploy" approved by HS, SA
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

**Comment by islean** [comment]
Test fails with following error:

```
sqlalchemy.exc.OperationalError: (pymysql.err.OperationalError) (3780, "Referencing column 'invoice_id' and referenced column 'id' in foreign key constraint 'pool_ibfk_3' are incompatible.")
[SQL: ALTER TABLE pool ADD FOREIGN KEY(invoice_id) REFERENCES invoice (id)]
(Background on this error at: http://sqlalche.me/e/13/e3q8)
```

**Comment by islean** [comment]
Tests fail with:

```
sqlalchemy.exc.OperationalError: (pymysql.err.OperationalError) (1091, "Can't DROP 'customer_invoice_pk'; check that column/key exists")
[SQL: 
DROP INDEX customer_invoice_pk ON invoice]
```

**Comment by islean** [comment]
Currently seeing a strange behaviour caused by there being 1467 samples with invoice_id = 0. There is no entry in the invoice table with id 0, rendering a foreign key from sample invoice impossible.

**Comment by islean** [comment]
Latest entry with invoice_id = 0 was created in 2018. Safest way to proceed might be to create a dummy entry in the invoice table with id 0 so that we can keep track of these samples. Once they have an invoice, we should be able to create the foreign key though.

**Comment by henrikstranneheim** [comment]
Sounds good as long as it easy to understand that this is a dummy entry.

**Comment by islean** [comment]
Added dummy invoice in stage:

image

**Comment by islean** [comment]
Still something wrong:

`[SQL: ALTER TABLE invoice MODIFY customer_id INTEGER NOT NULL]`

**Comment by islean** [comment]
Tying it to cust000 at least made the deployment work:

image

**Comment by islean** [comment]
Deployment to cg-vm1 went well and all tables are seemingly updated. E.g. sample.invoice_id now is a foreign key to invoice.id:

image

**Comment by islean** [comment]
Downgrade worked:

image

**Review by henrikstranneheim — APPROVED** [review]
LGTM, but this Alembic script does a lot. It would be great with some more eyes on it @Clinical-Genomics/sysdev

File: alembic/versions/936887038ba8_sync_cg_and_statusdb.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    op.alter_column(
+        "family",
+        "priority",
+        existing_type=mysql.ENUM("research", "standard", "priority", "express", "clinical_trials"),
+        nullable=False,
+    )
+    op.alter_column("invoice", "customer_id", existing_type=mysql.INTEGER(), nullable=False)
+    op.alter_column("invoice", "comment", existing_type=mysql.TEXT(), nullable=True)
+
+    op.drop_index("customer_invoice_pk", table_name="invoice")
```
Comment (seallard) [inline alembic/versions/936887038ba8_sync_cg_and_statusdb.py:36]: How come we are dropping this index?

**Review by seallard — APPROVED** [review]
Looks good 👍 
What if some of the values currently stored in the nullable columns are null? Maybe we need some default value or modify the data before it is possible to apply this migration in prod

File: alembic/versions/936887038ba8_sync_cg_and_statusdb.py (hunk 1 above)
Comment (islean) [reply alembic/versions/936887038ba8_sync_cg_and_statusdb.py:36]: It is basically a conversion with the line below. In the Python model we have this as a foreign key but in the actual database we have it as an index. Foreign keys are automatically indexed, though, so the line beneath should ensure that we have an index and a foreign key on the column. 

At the moment there would be no constraint to going into the db manually and an entry in the invoice with a customer_id which does not exist.

**Comment by islean** [comment]
> Looks good 👍 What if some of the values currently stored in the nullable columns are null? Maybe we need some default value or modify the data before it is possible to apply this migration in prod

I would be surprised given that the tests have been on a copy of production and I have tried to check it but I agree, this is a concern. It won't be that the columns with null will be dropped, but given that alembic currently assumes a non-transactional DDL it could be a bit cumbersome resetting a broken alembic deployment, since some but not all changes have gone through. This would break the downgrade script since the alembic version has not been updated in the db table, and even if it had been, it would assume all changes have been made.

What do we think - would it be enough to check manually before deployment that all those tables fulfill the non-nullable requirements or would we want to do a for-loop à la "set "dummy_value" where null"? @Clinical-Genomics/sysdev ?

**Comment by ChrOertlin** [comment]
> > Looks good 👍 What if some of the values currently stored in the nullable columns are null? Maybe we need some default value or modify the data before it is possible to apply this migration in prod
> 
> I would be surprised given that the tests have been on a copy of production and I have tried to check it but I agree, this is a concern. It won't be that the columns with null will be dropped, but given that alembic currently assumes a non-transactional DDL it could be a bit cumbersome resetting a broken alembic deployment, since some but not all changes have gone through. This would break the downgrade script since the alembic version has not been updated in the db table, and even if it had been, it would assume all changes have been made.
> 
> What do we think - would it be enough to check manually before deployment that all those tables fulfill the non-nullable requirements or would we want to do a for-loop à la "set "dummy_value" where null"? @Clinical-Genomics/sysdev ?

Can we get an overview that contains:
1. the extent to which these null values exist?
2. What are these attributes for which null values exist
3. Is there a way to fetch the data in question from some place else and fill in the proper values, if they exist?

Maybe this could be an attempt to make the database more complete?

**Comment by islean** [comment]
> > > Looks good 👍 What if some of the values currently stored in the nullable columns are null? Maybe we need some default value or modify the data before it is possible to apply this migration in prod
> > 
> > 
> > I would be surprised given that the tests have been on a copy of production and I have tried to check it but I agree, this is a concern. It won't be that the columns with null will be dropped, but given that alembic currently assumes a non-transactional DDL it could be a bit cumbersome resetting a broken alembic deployment, since some but not all changes have gone through. This would break the downgrade script since the alembic version has not been updated in the db table, and even if it had been, it would assume all changes have been made.
> > What do we think - would it be enough to check manually before deployment that all those tables fulfill the non-nullable requirements or would we want to do a for-loop à la "set "dummy_value" where null"? @Clinical-Genomics/sysdev ?
> 
> Can we get an overview that contains:
> 
>     1. the extent to which these null values exist?
> 
>     2. What are these attributes for which null values exist
> 
>     3. Is there a way to fetch the data in question from some place else and fill in the proper values, if they exist?
> 
> 
> Maybe this could be an attempt to make the database more complete?

I have double checked when I did it on stage and when going through the autorevision to see what changes should be done to it. Also, Iíf there had been any such entries, the deployment to stage would have failed. So the extent unless any new entries have been malformed is zero. And since there are restrictions in the Python models, anything added via the CLI or via the web interface will already have these restrictions.

**Comment by islean** [comment]
It would be more a question of "If we have missed anything - should we have a backup plan in the alembic script?" Of course this would have some problems, such as needing a decision on what to do with the values during the downgrade. 

I am in favour of believing that a deployment to stage immediately after copying prod to stage is enough of a security to forego this.

**Comment by islean** [comment]
Deployment successful after adding constraint on Python model make Analysis.family_id non-nullable.

image

**Comment by islean** [comment]
Deployment to cg-vm1 successful as well, although for some reason you need to rebuild the docker image sometimes in order to deploy it successfully.

**Comment by islean** [comment]
Downgrade successful: 

image

File: alembic/versions/936887038ba8_sync_cg_and_statusdb.py (hunk 1 above)
Comment (diitaz93) [reply alembic/versions/936887038ba8_sync_cg_and_statusdb.py:36]: I will resolve this conversation.

**Comment by diitaz93** [comment]
Deployed to stage, deployment failed in production due to an SQLAlchemy error

**Comment by diitaz93** [comment]
Deployment done in fix #2490

**Pull request merged**